user-guide: Adding rest api guide (#295)
foundriesio/docs: source/user-guide/foundriesio-rest-api/foundriesio-rest-api.rst

.. _ug-foundriesio-rest-api:

Foundries.io REST API
=====================

Foundries.io implements a `REST API`_ that allow users to interact with their 
FoundriesFactory and devices. This section shows examples of how to use 
the REST API with cURL_, a command-line tool for transferring data using 
various network protocols.

Creating Token
--------------

To interact with the `Foundries.io API <https://api.foundries.io/ota/>`_, you need to create a token.

Go to `Tokens <https://app.foundries.io/settings/tokens/>`_ and create a new **Api Token** by clicking on 
:guilabel:`+ New Token`.

Complete with a **Description** and the **Expiration date** and select :guilabel:`next`.

For this tutorial, check the :guilabel:`Use for tools like fioctl` box and 
select your **Factory**.

.. figure:: /_static/userguide/foundriesio-rest-api/token.png
   :width: 500
   :align: center

   Token for REST API

Save the output of this command to your copy buffer. This is your ``<API_TOKEN>`` value.

To follow this section, it is important to have:

> - Completed the :ref:`sec-learn` until the :ref:`gs-flash-device` section.

Reviewer: I would move these lines up above line 11

Getting Started
---------------

The `Foundries.io API <https://api.foundries.io/ota/>`_  includes a complete 
`OpenAPI <https://swagger.io/specification/>`_ definition. To get started, 
configure your host terminal with the variables below:

> Start by setting the variable ``FACTORY`` with your ``<FACTORY>`` name:
> 
> .. prompt:: bash host:~$, auto
> 
>     host:~$ export FACTORY=<FACTORY>
> 
> Set the variable ``TOKEN`` with the ``<API_TOKEN>`` you just created:
> 
> .. prompt:: bash host:~$, auto
> 
>     host:~$ export TOKEN=<API_TOKEN>

Reviewer: I would do this in a cleaner way, but it's just a suggestion
```suggestion
.. prompt:: bash host:~$, auto
    host:~$ export FACTORY=<FACTORY>
    host:~$ export TOKEN=<API_TOKEN>

..note ::
    Make sure to replace ``<FACTORY>`` with the factory name and ``<API_TOKEN>`` with the token created in the 
last step.
```

Device List
-----------

This function returns a list of devices registered in your Factory. The HTTP response includes 

> the same data you see in the :guilabel:`Devices` tab of the Factory. You can find this 

Reviewer: ```suggestion
the same data from the :guilabel:`Devices` tab of the Factory. You can find this 
```

> This function returns a list of devices registered in your Factory. The return is 
> the same you see in the :guilabel:`Devices` tab of the Factory. You can find this 
> resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_, 

Reviewer 2: ```suggestion
This function returns a list of devices registered in your Factory. The HTTP response includes 
the same data you see in the :guilabel:`Devices` tab of the Factory. You can find this 
resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_ 
```

> under devices.

Reviewer: ```suggestion
under _devices_.
```

          "owner": "5e13232f73927550af883e7b",
          "factory": "getting-started",
          "name": "device-01",
          "created-at": "2021-11-11T14:17:50+00:00",
          "last-seen": "2021-11-11T14:17:57+00:00",
          "ostree-hash": "cfacf42873aa06fbf53d7b1bfcb72032f21631b068620e575696ab2ab0670efd",
          "target-name": "raspberrypi3-64-lmp-2",
          "current-update": "",
          "device-tags": [
            "devel"
          ],
          "tag": "devel",
          "docker-apps": [],
          "network-info": {
            "hostname": "raspberrypi3-64",
            "local_ipv4": "192.168.15.13",
            "mac": "b8:27:eb:ca:78:75"
          },
          "up-to-date": true,
          "public-key": "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE2l3TY/Gu8W+euAA8bVufhxmCK0YR\ngt3B8XPXGS4rVRD3jfjjf7cCJQ4GLtvs+phTgeLiRJwTyWzsFZGYxqN1aA==\n-----END PUBLIC KEY-----\n",
          "is-prod": false,
          "is-wave": false
        }
      ],
      "page": 1,
          "pages": 1,
          "limit": 1000,
      "total": 1
    }

Targets List
------------

This function returns a list of your Factory targets. The HTTP response includes  

> the same data you see in the :guilabel:`Targets` tab of the Factory. You can find this 

Reviewer: ```suggestion
the same data from the :guilabel:`Targets` tab of the Factory. You can find this 
```

> This function returns a list of your Factory targets. The return is 
> the same you see in the :guilabel:`Targets` tab of the Factory. You can find this 
> resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_, under factories.

Reviewer 2: ```suggestion
This function returns a list of your Factory targets. The HTTP response includes 
the same data you see in the :guilabel:`Targets` tab of the Factory. You can find this 
resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_ under factories.
```

> resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_ under factories.

Reviewer: ```suggestion
resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_ under _factories_.
```

           "version": "3"
         },
         "hashes": {
           "sha256": "f190733c576920a203acd40d8df1cb82135aaf50f0b2c5fa8402745d73a633b3"
         },
         "length": 0
       },
       "raspberrypi3-64-lmp-2": {
         "custom": {
           "arch": "aarch64",
           "cliUploaded": false,
           "createdAt": "2021-07-28T20:15:29Z",
           "hardwareIds": [
             "raspberrypi3-64"
           ],
           "image-file": "lmp-factory-image-raspberrypi3-64.wic.gz",
           "lmp-manifest-sha": "b6483a7b0bd666b5b871662fa46477cdeede80f2",
           "meta-subscriber-overrides-sha": "7de1123998c9b362df278132fde8fccb57215647",
           "name": "raspberrypi3-64-lmp",
           "tags": [
             "devel"
           ],
           "targetFormat": "OSTREE",
           "updatedAt": "2021-07-28T20:15:29Z",
           "uri": "https://ci.foundries.io/projects/getting-started/lmp/builds/2",
           "version": "2"
         },
         "hashes": {
           "sha256": "cfacf42873aa06fbf53d7b1bfcb72032f21631b068620e575696ab2ab0670efd"
         },
         "length": 0
       }
     }

Configure Device
----------------

> It's also possible to send configuration files to your device or fleet. 

Reviewer: ```suggestion
It is also possible to send configuration files to your device or fleet. 
```

You can find more about it in the tutorial section: :ref:`tutorial-dynamic-configuration-file`.

Assuming you want to send a configuration file named as ``app.config`` to your device:

**app.config**:

.. prompt:: text

     [page]
     url = "https://foundries.io/"

First, you have to convert the file contents into properly escaped JSON string:

.. prompt:: text

>      [page]\nurl = \"https://foundries.io/\"\n"}

Reviewer 2: ```suggestion
     [page]\nurl = \"https://foundries.io/\"\n"
```

Author: Sent another Fixup @doanac

Now use the command below to create the ``API_DATA`` variable.

> Note that the file content is on ``value`` and the file name on the ``name``.

Reviewer: ```suggestion
Note that the file content is on ``value`` and the file name on ``name``.
```

.. prompt:: bash host:~$, auto

    host:~$ export API_DATA='{ "reason": "API test", "files": [{"name":"app.config","unencrypted":true,"value":"[page]\nurl = \"https://foundries.io/\"\n"}]}'

> Define the variable ``DEVICE_NAME`` and using cURL_ with ``-x PATCH``, configure your device:

Reviewer: ```suggestion
Define the variable ``DEVICE_NAME`` and configure your device using cURL_ with ``-x PATCH``:
```

Reviewer: should it be -x or -X? the command below shows -X

.. prompt:: bash host:~$, auto

    host:~$ export DEVICE_NAME=<DEVICE_NAME>
    host:~$ curl --data "$API_DATA" -H "Content-Type: application/json" -H "OSF-TOKEN: $TOKEN" -X PATCH https://api.foundries.io/ota/devices/$DEVICE_NAME/config/

**Example Output**:

.. prompt:: text

     {
       "created-at": "2021-11-11T15:59:07",
       "applied-at": null,
       "reason": "API test",
       "files": [
         {
           "name": "app.config",
           "value": "[page]\nurl = \"https://foundries.io/\"\n",
           "unencrypted": true
         }
       ]
     }

> After some time, ``fioconfig`` receives all the changes and you can find your file 
> on your device at ``/var/run/secrets/``.

Reviewer: ```suggestion
After some time, ``fioconfig`` receives all the changes and you can find the configuration file 
on your device at ``/var/run/secrets/``.
```

.. prompt:: bash device:~$, auto

    device:~$ sudo cat /var/run/secrets/app.config

**Example Output**:

.. prompt:: text

     [page]
     url = "https://foundries.io/"

Enabling/Disabling Application
------------------------------

It is also possible to configure what application or tag the device should use.
For example, to enable the ``shellhttpd`` application, you have to send a JSON file similar to:

.. prompt:: text

    {
      "name": "z-50-fioctl.toml",
      "on-changed": [
        "/usr/share/fioconfig/handlers/aktualizr-toml-update"
      ],
      "unencrypted": true,
          "value": "\n[pacman]\n  compose_apps = \"shellhttpd\"\n"
    }

The command below adds the json file into the ``API_DATA`` variable:

.. prompt:: bash host:~$, auto

    host:~$ export API_DATA='{ "reason": "API test", "files": [{"name":"z-50-fioctl.toml","on-changed":["/usr/share/fioconfig/handlers/aktualizr-toml-update"],"unencrypted":true,"value":"\n[pacman]\n  compose_apps = \"shellhttpd\"\n"}]}'

> Define the variable ``DEVICE_NAME`` and using cURL_ with ``-x PATCH``, configure your device:

Reviewer: same as above (-X comment)
```suggestion
Define the variable ``DEVICE_NAME`` configure your device using cURL_ with ``-x PATCH``:
```

.. prompt:: bash host:~$, auto

    host:~$ export DEVICE_NAME=<DEVICE_NAME>
    host:~$ curl --data "$API_DATA" -H "Content-Type: application/json" -H "OSF-TOKEN: $TOKEN" -X PATCH https://api.foundries.io/ota/devices/$DEVICE_NAME/config/

**Example Output**:

.. prompt:: text

     {"created-at": "2021-11-10T19:02:30", "applied-at": null, "reason": "API test", "files": [{"name": "wireguard-client", "value": "enabled=0\n\npubkey=J0H7CMG10TsTEai2Ui35KV0fb5oaJ8qd+mnWgIu091s=", "unencrypted": true}, {"name": "z-50-fioctl.toml", "on-changed": ["/usr/share/fioconfig/handlers/aktualizr-toml-update"], "value": "\n[pacman]\n  compose_apps = \"shellhttpd\"\n", "unencrypted": true}]}

Finding More about REST API
---------------------------

Once logged into http://app.foundries.io, you may interact and play with the REST API 
from your browser at http://api.foundries.io/ota. 

> 1 - In the right top corner of the operation, you can find :guilabel:`Try it out`.
> 
> 2 - That allows you to customize parameters.

Reviewer: ```suggestion
1 - Click on the operation to test. That shows the :guilabel:`Try it out` option. Click on it.

2 - Customize parameters as you like.
```

.. _ug-foundriesio-rest-api:

Foundries.io REST API
=====================

For those who are not familiar with REST API:

`Representational state transfer`_ (REST) is a software architectural style that 
was created to guide the design and development of the architecture for the World 
Wide Web. (Wikipedia)

Foundries.io implements a REST API to allow users to interact with their 
FoundriesFactory and devices connected to it according to their needs.

This section shows examples of how to use the REST API with cURL_, a command-line 

> tool for transferring data using various network protocols.

Reviewer 2: i'm sorry. my suggestion has a problem. I think this would read better:
"Foundries.io implements a REST API **that** allow users to interact "

Creating Token
--------------

To interact with the `Foundries.io API <https://api.foundries.io/ota/>`_, you need to create a token.

Go to `Tokens <https://app.foundries.io/settings/tokens/>`_ and create a new **Api Token** by clicking on 
:guilabel:`+ New Token`.

Complete with a **Description** and the **Expiration date** and select :guilabel:`next`.

For this tutorial, check the :guilabel:`Use for tools like fioctl` box and 
select your **Factory**.

.. figure:: /_static/userguide/foundriesio-rest-api/token.png
   :width: 500
   :align: center

   Token for REST API

Save the output of this command to your copy buffer. This is your ``<API_TOKEN>`` value.

To follow this section, it is important to have:

- Completed the :ref:`sec-learn` until the :ref:`gs-flash-device` section.

Getting Started
---------------

> The `Foundries.io API <https://api.foundries.io/ota/>`_ can perform multiple operations. You can find all operations 
> with their corresponding functions and parameters at `Foundries.io API <https://api.foundries.io/ota/>`_.
> 
> Configure your host terminal with the variable below:

Reviewer 2: ```suggestion
```suggestion
The `Foundries.io API <https://api.foundries.io/ota/>`_  includes a complete `OpenAPI <https://swagger.io/specification/>`_ definition. To get started, configure your host terminal with the variables below:
```

Start by setting the variable ``FACTORY`` with your ``<FACTORY>`` name:

.. prompt:: bash host:~$, auto

    host:~$ export FACTORY=<FACTORY>

Set the variable ``TOKEN`` with the ``<API_TOKEN>`` you just created:

.. prompt:: bash host:~$, auto

    host:~$ export TOKEN=<API_TOKEN>

Device List
-----------

This function returns a list of devices registered in your Factory. The return is 
the same you see in the :guilabel:`Devices` tab of the Factory.

> You can find this function at `Foundries.io API <https://api.foundries.io/ota/>`_, under devices.

Reviewer 2: These are called "resources" in the REST world.

```suggestion
You can find this resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_, under devices.
```

Reviewer 2: You should probably also make these two paragraphs a single paragraph.

> Using the terminal, run the cURL_ command below with ``-x GET`` to get the devices list:
> 
> .. prompt:: bash host:~$, auto
> 
>     host:~$ curl -H "OSF-TOKEN: $TOKEN" -X GET https://api.foundries.io/ota/devices/

Reviewer 2: "-X GET" isn't required for curl

```suggestion
Using the terminal, run:

.. prompt:: bash host:~$, auto

    host:~$ curl -H "OSF-TOKEN: $TOKEN" https://api.foundries.io/ota/devices/
```

          "factory": "getting-started",
          "name": "device-01",
          "created-at": "2021-11-11T14:17:50+00:00",
          "last-seen": "2021-11-11T14:17:57+00:00",
          "ostree-hash": "cfacf42873aa06fbf53d7b1bfcb72032f21631b068620e575696ab2ab0670efd",
          "target-name": "raspberrypi3-64-lmp-2",
          "current-update": "",
          "device-tags": [
            "devel"
          ],
          "tag": "devel",
          "docker-apps": [],
          "network-info": {
            "hostname": "raspberrypi3-64",
            "local_ipv4": "192.168.15.13",
            "mac": "b8:27:eb:ca:78:75"
          },
          "up-to-date": true,
          "public-key": "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE2l3TY/Gu8W+euAA8bVufhxmCK0YR\ngt3B8XPXGS4rVRD3jfjjf7cCJQ4GLtvs+phTgeLiRJwTyWzsFZGYxqN1aA==\n-----END PUBLIC KEY-----\n",
          "is-prod": false,
          "is-wave": false
        }
      ],
      "page": 1,
          "pages": 1,
          "limit": 1000,
      "total": 1
    }

Targets List
------------

This function returns a list of your Factory targets. The return is 
the same you see in the :guilabel:`Targets` tab of the Factory.

> You can find this function at `Foundries.io API <https://api.foundries.io/ota/>`_, under factories:

Reviewer 2: ```suggestion
You can find this resource's documentation at `Foundries.io API <https://api.foundries.io/ota/>`_, under factories.
```

> Using the terminal, run the cURL_ command below with ``-x GET`` to get the devices list:
> 
> .. prompt:: bash host:~$, auto
> 
>     host:~$ curl -H "OSF-TOKEN: $TOKEN" -X GET https://api.foundries.io/ota/factories/$FACTORY/targets/
>  

Reviewer 2: ```suggestion
Using the terminal, run:

.. prompt:: bash host:~$, auto

    host:~$ curl -H "OSF-TOKEN: $TOKEN" https://api.foundries.io/ota/factories/$FACTORY/targets/
 
```

           "version": "3"
         },
         "hashes": {
           "sha256": "f190733c576920a203acd40d8df1cb82135aaf50f0b2c5fa8402745d73a633b3"
         },
         "length": 0
       },
       "raspberrypi3-64-lmp-2": {
         "custom": {
           "arch": "aarch64",
           "cliUploaded": false,
           "createdAt": "2021-07-28T20:15:29Z",
           "hardwareIds": [
             "raspberrypi3-64"
           ],
           "image-file": "lmp-factory-image-raspberrypi3-64.wic.gz",
           "lmp-manifest-sha": "b6483a7b0bd666b5b871662fa46477cdeede80f2",
           "meta-subscriber-overrides-sha": "7de1123998c9b362df278132fde8fccb57215647",
           "name": "raspberrypi3-64-lmp",
           "tags": [
             "devel"
           ],
           "targetFormat": "OSTREE",
           "updatedAt": "2021-07-28T20:15:29Z",
           "uri": "https://ci.foundries.io/projects/getting-started/lmp/builds/2",
           "version": "2"
         },
         "hashes": {
           "sha256": "cfacf42873aa06fbf53d7b1bfcb72032f21631b068620e575696ab2ab0670efd"
         },
         "length": 0
       }
     }

Configure Device
----------------

> With `Foundries.io API <https://api.foundries.io/ota/>`_, it is possible to send 
> configuration files to your device or fleet. You can find more about it in the 
> tutorial section: :ref:`tutorial-dynamic-configuration-file`.

Reviewer 2: ```suggestion
It's also possible to send 
configuration files to your device or fleet. You can find more about it in the 
tutorial section: :ref:`tutorial-dynamic-configuration-file`.
```

Assuming you want to send a configuration file named as ``app.config`` to your device:

**app.config**:

.. prompt:: text

     [page]
     url = "https://foundries.io/"

> First, you have to convert it to a inline content:

Reviewer 2: ```suggestion
First, you have to convert the file contents into properly escaped JSON string:
```

    }

The command below adds the json file into the ``API_DATA`` variable:

.. prompt:: bash host:~$, auto

    host:~$ export API_DATA='{ "reason": "API test", "files": [{"name":"z-50-fioctl.toml","on-changed":["/usr/share/fioconfig/handlers/aktualizr-toml-update"],"unencrypted":true,"value":"\n[pacman]\n  compose_apps = \"shellhttpd\"\n"}]}'

Define the variable ``DEVICE_NAME`` and using cURL_ with ``-x PATCH``, configure your device:

.. prompt:: bash host:~$, auto

    host:~$ export DEVICE_NAME=<DEVICE_NAME>
    host:~$ curl --data "$API_DATA" -H "Content-Type: application/json" -H "OSF-TOKEN: $TOKEN" -X PATCH https://api.foundries.io/ota/devices/$DEVICE_NAME/config/

**Example Output**:

.. prompt:: text

     {"created-at": "2021-11-10T19:02:30", "applied-at": null, "reason": "API test", "files": [{"name": "wireguard-client", "value": "enabled=0\n\npubkey=J0H7CMG10TsTEai2Ui35KV0fb5oaJ8qd+mnWgIu091s=", "unencrypted": true}, {"name": "z-50-fioctl.toml", "on-changed": ["/usr/share/fioconfig/handlers/aktualizr-toml-update"], "value": "\n[pacman]\n  compose_apps = \"shellhttpd\"\n", "unencrypted": true}]}

Finding More about REST API
---------------------------

> When you access the `Foundries.io API <https://api.foundries.io/ota/>`_, page, 
> you can test the REST API from there.
> 
> First, you have to make sure you are logged in to your `Foundries.io API <https://api.foundries.io/ota/>`_. 
> That will automatically give you access to your Factory.
> 
> Go to `Foundries.io API <https://api.foundries.io/ota/>`_, and choose one of the operations.

Reviewer 2: ```suggestion
Once logged into app.foundries.io, you may interact and play with the REST API from your browser at http://api.foundries.io/ota. 
```

Author: @doanac done!

> .. _ug-foundriesio-rest-api:

Reviewer 2: A general comment that I couldn't attach to all the images. I worry we are creating a ton of technical debt with all the images in places that don't absolutely require them.

The webui will evolve and these will all become misleading or inaccurate. Someone's going to have to comb through dozens and dozens of these and re-create accurate ones. In some cases they might have to re-generate similar factory data to produce the right view?

so something like `response_api_example.png` - maybe that could be preformatted text?

Author: I removed most of the images to make it simple.

    imx8mmevk-brlab-10        demo     imx8mmevk-lmp-825                OK                                  debian-base_flutter_foundries  true        false
    intel-corei7-64-brlab-16  demo     intel-corei7-64-lmp-809          OK                                                                 true        false
    raspberrypi3-64-brlab-11  demo     raspberrypi3-64-lmp-810          OK                                                                 true        false
    raspberrypi3-64-brlab-13  demo     raspberrypi3-64-lmp-786          OFFLINE                                                            false       false
    raspberrypi4-64-brlab-12  demo     raspberrypi4-64-lmp-810          OK                                                                 true        false
    raspberrypi4-64-brlab-14  demo     raspberrypi4-64-lmp-810          OK                                  gin,gin_ui                     true        false
    stm32mp1-disco-brlab-17   demo     stm32mp1-disco-lmp-810           OK                                                                 true        false

Of course, both methods don’t scale for software development and this is 

> where the REST API could help you access and configure what you need directly.

Reviewer 2: I think you can drop all of this. We should step back and decide the target audience. If its someone building a web page like you state, they know what REST is already. 

I'm also not sure its fair to say our current tooling "doesn't scale with development". I think you just put a hyperlink to REST below and you've got a way for people to click on a information about REST.

Author: I removed most of the images to make them simple.

 
**Example Output**:

.. prompt:: text

       "devices": [
         {
           "uuid": "91482180-6be2-4aa8-8180-75d30fdabd57",
           "owner": "5e13232f73927550af883e7b",
           "factory": "demo",
           "name": "imx8mmevk-brlab-10",
           "created-at": "2021-07-30T22:05:09+00:00",
           "last-seen": "2021-11-10T18:30:00+00:00",
           "ostree-hash": "6635beeefd577c4471d0a6eeacc3d2c836ac1d9b3d23c42e731b76b956cd2f61",
           "target-name": "imx8mmevk-lmp-825",
           "current-update": "",
           "device-tags": [
             "flutter"
           ],
           "tag": "flutter",
           "docker-apps": [
             "debian-base_flutter_foundries"
           ],
           "network-info": {
             "hostname": "imx8mmevk-brlab-10",
             "local_ipv4": "192.168.15.10",
             "mac": "00:04:9f:06:40:c7"
           },
           "up-to-date": true,
           "public-key": "-----BEGIN PUBLIC KEY-----\nMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAETSZO1PVC4Q7FTLKyHCb55X2U4Wrf\nJNIPcG5+aFlTQ7LkDvwtu8RypaLSUJp6/as3s+T/2uRAsDTZ902IH1BccQ==\n-----END PUBLIC KEY-----\n",
           "is-prod": false,
           "is-wave": false

>          },

Reviewer 2: this is pretty excessive, maybe just show 1 or 2 devices and then a "..."

         },
         "length": 0
       },
       "raspberrypi3-64-lmp-2": {
         "custom": {
           "arch": "aarch64",
           "cliUploaded": false,
           "createdAt": "2021-07-28T20:15:29Z",
           "hardwareIds": [
             "raspberrypi3-64"
           ],
           "image-file": "lmp-factory-image-raspberrypi3-64.wic.gz",
           "lmp-manifest-sha": "b6483a7b0bd666b5b871662fa46477cdeede80f2",
           "meta-subscriber-overrides-sha": "7de1123998c9b362df278132fde8fccb57215647",
           "name": "raspberrypi3-64-lmp",
           "tags": [
             "devel"
           ],
           "targetFormat": "OSTREE",
           "updatedAt": "2021-07-28T20:15:29Z",
           "uri": "https://ci.foundries.io/projects/getting-started/lmp/builds/2",
           "version": "2"
         },
         "hashes": {
           "sha256": "cfacf42873aa06fbf53d7b1bfcb72032f21631b068620e575696ab2ab0670efd"
         },
         "length": 0
       }
     }

Configure Device
----------------

With Foundries.io REST API it is possible to configure or update device configuration.
For example, to select the ``shellhttpd`` application, you have to send a JSON file similar to:

> 

Reviewer 2: How about we just show how to configure a generic file on the device? that's probably a more common thing than "enabling shellhttpd".

Author: This was one of the customer questions that motivated me to write about it.

Author: Maybe we should add another example on how to configure the device, but not I don't think remove enable app would be a good idea.

   Devices on FoundriesFactories API

Next, click on :guilabel:`Try it out`.

As you can see, it allows you to customize the operation. Next, click on :guilabel:`Execute`:

.. figure:: /_static/userguide/foundriesio-rest-api/customize_api_example.png
   :width: 900
   :align: center

   Customizing Operation on FoundriesFactories API

It gives you a cURL_ command reference:

.. figure:: /_static/userguide/foundriesio-rest-api/curl_api_example.png
   :width: 900
   :align: center

   cURL reference on FoundriesFactories API

And finally the command response:

.. figure:: /_static/userguide/foundriesio-rest-api/response_api_example.png
   :width: 900
   :align: center

   Response of the try it out on FoundriesFactories API

This is the best way to explore the Foudries.io REST API and all its possibilities.

Python Example
--------------

Following it is a very simple example of how to get the devices list using Python :

.. prompt:: bash host:~$, auto

    host:~$ python3
     Python 3.8.10 (default, May  4 2021, 00:00:00) 
     [GCC 10.2.1 20201125 (Red Hat 10.2.1-9)] on linux
     Type "help", "copyright", "credits" or "license" for more information.
     >>> import requests

>      >>> token = "sSDjXx8t4MVs7h6f6j9vrZSajXBEJQckijogPpM6"

Reviewer 2: you've leaked a token. delete it.

     >>> headers = {"OSF-TOKEN": token}
     >>> r = requests.get("https://api.foundries.io/ota/devices/", headers=headers)
     >>> import json

>      >>> print(json.dumps(r.json(), indent=2))

Reviewer 2: i don't think we need a python example. `requests` isn't in the python standard library, so then you wind up going down a rabbit hole of requirements and such.
